File: C/examples/large_example.c
#include "textformats_c.h"
#include <stdio.h>
#include <assert.h>

void decoded_processor(JsonNode* node, void* data)
{
  printf("Decoded value: %s\n", jsonnode_to_string(node));
}

int main(void)
{
  NimMain();
  Specification *fas_spec;
  DatatypeDefinition *fas_entry, *fas_header;
  JsonNode *node;

  char *header = ">ABCD some sequence";
  char *decoded_header_json =
    "{\"fastaid\":\"ABCD\",\"desc\":\"some sequence\"}";
  printf("Encoded: %s\n", header);
  fas_spec = tf_specification_from_file("../../spec/fasta.yaml");
  if (!tf_is_compiled("../../spec/fasta.yaml"))
    printf("Spec fasta.yaml is not compiled");
  else assert(false);
  tf_compile_specification("../../spec/fasta.yaml", "fasta.tfs");
  if (tf_is_compiled("fasta.tfs"))
    printf("Spec fasta.tfs is compiled");
  else assert(false);
  tf_run_specification_testfile(fas_spec, "../../spec/fasta.yaml");
  fas_entry = tf_default_definition(fas_spec);
  fas_header = tf_get_definition(fas_spec, "header");
  printf("%s\n", tf_describe(fas_header));
  printf("Datatypes: %s\n", tf_datatype_names(fas_spec));
  node = tf_decode(header, fas_header);
  printf("%s\n", jsonnode_to_string(node));
  delete_jsonnode(node);
  printf("%s\n", tf_decode_to_json(header, fas_header));
  if (tf_is_valid_encoded(header, fas_header))
    printf("%s is_valid\n", header);
  else assert(false);
  printf("%s\n", tf_encode_json(decoded_header_json, fas_header));
  node = new_j_object();
  j_object_add(node, "fastaid", new_j_string("ABCD"));
  j_object_add(node, "desc", new_j_string("some sequence"));
  printf("%s\n", tf_encode(node, fas_header));
  if (tf_is_valid_decoded(node, fas_header))
    printf("decoded is_valid\n");
  else assert(false);
  delete_jsonnode(node);
  if (tf_is_valid_decoded_json(decoded_header_json, fas_header))
    printf("%s is_valid\n", decoded_header_json);
  else assert(false);
  tf_decode_file("../../tests/testdata/bio/test.fas", false, fas_entry,
                 decoded_processor, NULL, false);
  tf_delete_definition(fas_entry);
  tf_delete_definition(fas_header);
  tf_delete_specification(fas_spec);
  return 0;
}

File: python/examples/large_example.py
#!/usr/bin/env python3
import os
from textformats import Specification, DECODED_PROCESSOR_LEVEL

testdir = os.path.dirname(os.path.realpath(__file__))
specdir = testdir + "/../../spec"
biotestdir = testdir + "/../../tests/testdata/bio"

header = ">ABCD some sequence"
decoded_header = {"fastaid": "ABCD", "desc": "some sequence"}
decoded_header_json = "{\"fastaid\":\"ABCD\",\"desc\":\"some sequence\"}";

print(f"Encoded: {header}")
spec = Specification(f"{specdir}/fasta.yaml")
assert(not spec.is_compiled)
print("Spec fasta.yaml is not compiled")
spec.compile(f"{specdir}/fasta.yaml", "fasta.tfs")
spec = Specification("fasta.tfs")
assert(spec.is_compiled)
print("Spec fasta.tfs is compiled")
spec.test(f"{specdir}/fasta.yaml")

fas_entry = spec["default"]
fas_header = spec["header"]
print(fas_header)
print(spec.datatype_names)

print(fas_header.decode(header))
print(fas_header.decode(header, True))
assert(fas_header.is_valid_encoded(header))
print(f"\"{header}\" is a valid encoded fas_header")
print(fas_header.encode(decoded_header_json, True))
print(fas_header.encode(decoded_header))
assert(fas_header.is_valid_decoded(decoded_header))
print(f"\"{decoded_header_json}\" is a valid decoded fas_header")

def decoded_processor(decoded, processor_data):
  print(f"Decoded value: {decoded}")

print("\nDecode file, level \"whole\"")
fas_entry.decode_file(f"{biotestdir}/test.fas", decoded_processor,
    None, DECODED_PROCESSOR_LEVEL.WHOLE)
print("\nDecode file, level \"element\"")
fas_entry.decode_file(f"{biotestdir}/test.fas", decoded_processor,
    None, DECODED_PROCESSOR_LEVEL.ELEMENT)
print("\nDecode file, level \"line\"")
fas_entry.decode_file(f"{biotestdir}/test.fas", decoded_processor,
    None, DECODED_PROCESSOR_LEVEL.LINE)
